Layout/editable docs for 2.0 (#2942)
conan-io/docs: reference/conanfile/methods/layout.rst

.. _reference_conanfile_methods_layout:

layout()
========

In the layout() method you can adjust ``self.folders`` and ``self.cpp``.

.. _layout_folders_reference:

self.folders
^^^^^^^^^^^^

- **self.folders.source** (Defaulted to ""): Specifies a subfolder where the sources are.
  The ``self.source_folder`` attribute inside the ``source(self)`` and ``build(self)``
  methods will be set with this subfolder. The *current working directory* in the
  ``source(self)`` method will include this subfolder. The `export_sources`, `exports` and

>   `scm` sources will also be copied to the root source directory. It is used in the cache

Author: Thanks, that was a mistake

>   ``source(self)`` method will include this subfolder. The `export_sources`, `exports` and
>   `scm` sources will also be copied to the root source directory. It is used in the cache

Author: ```suggestion
  ``source(self)`` method will include this subfolder. The `export_sources` and `exports` sources will also be copied to the root source directory. It is used in the cache
```

  when running :command:`conan create` (relative to the cache source folder) as well as in
  a local folder when running :command:`conan build` (relative to the local current
  folder).

- **self.folders.build** (Defaulted to ""): Specifies a subfolder where the files from the
  build are. The ``self.build_folder`` attribute and the *current working directory*
  inside the ``build(self)`` method will be set with this subfolder. It is used in the
  cache when running :command:`conan create` (relative to the cache source folder) as well
  as in a local folder when running :command:`conan build` (relative to the local current
  folder).

> - **self.folders.generators** (Defaulted to ""): Specifies a subfolder where to write the

Reviewer: ```suggestion
- **self.folders.generators** (Defaulted to ``""``): Specifies a subfolder in which to write the
```

>   files from the generators and the toolchains. In the cache, when running the

Reviewer: ```suggestion
  files from the generators and the toolchains. In the cache, when running
```

  :command:`conan create`, this subfolder will be relative to the root build folder and
  when running the :command:`conan install` command it will be relative to the current
  working directory.

> - **self.folders.root** (Defaulted to None): Specifies a parent directory where the

Reviewer: ```suggestion
- **self.folders.root** (Defaulted to ``None``): Specifies a parent directory where the
```

  sources, generators, etc., are located specifically when the ``conanfile.py`` is located
  in a separated subdirectory. Check :ref:`this
  example<examples_conanfile_layout_conanfile_in_subfolder>` on how to use
  **self.folders.root**.

> - **self.folders.subproject** (Defaulted to None): Specifies a subfolder where the

Reviewer: ```suggestion
- **self.folders.subproject** (Defaulted to ``None``): Specifies a subfolder where the
```

  ``conanfile.py`` is relative to the project root. This is particularly useful for
  :ref:`layouts with multiple subprojects<examples_conanfile_layout_multiple_subprojects>`

.. _layout_cpp_reference:

self.cpp
^^^^^^^^

The ``layout()`` method allows to declare ``cpp_info`` objects not only for the final
package (like the classic approach with the ``self.cpp_info`` in the
``package_info(self)`` method) but for the ``self.source_folder`` and
``self.build_folder``.

> The fields of the cpp_info objects at ``self.cpp.build`` and ``self.cpp.source`` are the

Reviewer: ```suggestion
The fields of the ``cpp_info`` objects at ``self.cpp.build`` and ``self.cpp.source`` are the
```
